src/VecSim/algorithms/hnsw/hnswlib.h

```diff
@@ -405,6 +429,54 @@ dist_t HierarchicalNSW<dist_t>::processCandidate(tableint curNodeId, const void
     return lowerBound;
 }
 
+template <typename dist_t>
+void HierarchicalNSW<dist_t>::processCandidate_RangeSearch(tableint curNodeId,
+                                                           const void *query_data, size_t layer,
+                                                           float epsilon, tag_t visited_tag,
+                                                           VecSimQueryResult **results,
+                                                           candidatesMaxHeap<dist_t> &candidate_set,
+                                                           dist_t dyn_range, float radius) const {
+
+#ifdef ENABLE_PARALLELIZATION
+    std::unique_lock<std::mutex> lock(link_list_locks_[curNodeId]);
+#endif
+    linklistsizeint *node_ll = get_linklist_at_level(curNodeId, layer);
+    size_t links_num = getListCount(node_ll);
+    auto *node_links = (tableint *)(node_ll + 1);
+
+    __builtin_prefetch(visited_nodes_handler->getElementsTags() + *(node_ll + 1));
+    __builtin_prefetch(visited_nodes_handler->getElementsTags() + *(node_ll + 1) + 64);
+    __builtin_prefetch(getDataByInternalId(*node_links));
+    __builtin_prefetch(getDataByInternalId(*(node_links + 1)));
+
+    for (size_t j = 0; j < links_num; j++) {
+        tableint candidate_id = *(node_links + j);
+
+        __builtin_prefetch(visited_nodes_handler->getElementsTags() + *(node_links + j + 1));
+        __builtin_prefetch(getDataByInternalId(*(node_links + j + 1)));
+
+        if (this->visited_nodes_handler->getNodeTag(candidate_id) == visited_tag)
+            continue;
+        this->visited_nodes_handler->tagNode(candidate_id, visited_tag);
+        char *candidate_data = (getDataByInternalId(candidate_id));
+
+        dist_t candidate_dist = fstdistfunc_(query_data, candidate_data, dist_func_param_);
+        if (candidate_dist < dyn_range * (1.0f + epsilon)) {
+            candidate_set.emplace(-candidate_dist, candidate_id);
+
+            __builtin_prefetch(getDataByInternalId(candidate_set.top().second));
```

Comment:
We should understand why we need this. Maybe add an issue to the backlog

Reply:
I agree... perhaps I'll open an issue for Yuri